src/odemis/driver/npmc.py

```diff
@@ -635,6 +717,56 @@ def _createMoveFuture(self):
         f.task_canceller = self._cancelCurrentMove
         return f
 
+    @isasync
+    def reference(self, axes):
+        if not axes:
+            return model.InstantaneousFuture()
+        f = self._createMoveFuture()
+        f = self._executor.submitf(f, self._doReference, f, axes)
+        return f
+
+    def _doReference(self, future, axes):
+        """
+        Actually runs the referencing code
+        axes (set of str)
+        raise:
+            IOError: if referencing failed due to hardware
+            CancelledError if was cancelled
+        """
+        # Reset reference so that if it fails, it states the axes are not
+        # referenced (anymore)
+        with future._moving_lock:
+            try:
+                # do the referencing for each axis sequentially
+                # (because each referencing is synchronous)
+                # TODO: with the "standard" referencing, we could run them
+                # simultaneously
+                for a in axes:
+                    if future._must_stop.is_set():
+                        raise CancelledError()
+                    aid = self._axis_map[a]
+                    future._current_axis = aid
+                    self.referenced._value[a] = False
+                    self.HomeSearch(aid, REF_NEGATIVE_LIMIT)  # search for the negative limit signal to set an origin
+                    self._waitEndMove(future, (aid,), time.time() + 100)  # block until it's over
+                    self.SetHome(aid, 0.0)  # set negative limit as origin
+                    self.SaveMemory()  # save config on device
```

Comment:
Why do you have to save the reference position on the device?! Isn't that gone anyway after rebooting the device, as it doesn't know whether the axis has moved or not since last referencing? IOW, what would happen if you didn't save to memory?

Reply:
Theoretically, this function is supposed to save to the non-voltaile memory on the device so that it would retain the settings once it is turned off. But it doesn't seem to work when I tested it - maybe a CMOS battery is needed or something? Either way, I can remove it.